src/types/redis_hash.cc

```diff
@@ -426,4 +484,274 @@ rocksdb::Status Hash::RandField(engine::Context &ctx, const Slice &user_key, int
+
+  ret->reserve(fields.size());
+  auto now = util::GetTimeStampMS();
+  for (size_t i = 0; i < keys.size(); i++) {
+    if (!statuses[i].ok() && !statuses[i].IsNotFound()) return statuses[i];
+    auto value = values[i].ToString();
+    auto status = statuses[i];
+
+    if (status.IsNotFound()) {
+      ret->emplace_back(-2);
+      continue;
+    }
+
+    uint64_t expire = 0;
+    status = decodeExpireFromValue(metadata, &value, expire);
+    if (status.IsNotFound()) {
+      ret->emplace_back(-2);
+    } else if (expire == 0) {
+      ret->emplace_back(-1);
+    } else {
+      ret->emplace_back(int64_t(expire - now));
+    }
+  }
+  return rocksdb::Status::OK();
+}
+
+bool Hash::IsFieldExpired(const Slice &metadata_key, const Slice &value) {
+  HashMetadata hash_metadata(false);
+  if (!hash_metadata.Decode(metadata_key).ok()) {
+    return false;
+  }
+  if (!hash_metadata.IsFieldExpirationEnabled()) {
+    return false;
+  }
+  uint64_t expire = 0;
+  rocksdb::Slice data(value);
+  GetFixed64(&data, &expire);
+  return expire != 0 && expire < util::GetTimeStampMS();
+}
+
+uint64_t Hash::GetValidFieldCount(engine::Context &ctx, const Slice &ns_key, const HashMetadata &metadata) {
+  if (metadata.Expired()) {
+    return 0;
+  }
+  // if field expiration is disabled,
+  // the size field in metadata is the length of hash
+  if (!metadata.IsFieldExpirationEnabled()) {
+    return metadata.size;
+  }
+
+  std::string prefix_key = InternalKey(ns_key, "", metadata.version, storage_->IsSlotIdEncoded()).Encode();
+  std::string next_version_prefix_key =
+      InternalKey(ns_key, "", metadata.version + 1, storage_->IsSlotIdEncoded()).Encode();
+
+  rocksdb::ReadOptions read_options = ctx.DefaultScanOptions();
+  rocksdb::Slice upper_bound(next_version_prefix_key);
+  read_options.iterate_upper_bound = &upper_bound;
+
+  // count the num of expired field
+  uint64_t expired_field_num = 0;
+  auto iter = util::UniqueIterator(ctx, read_options);
+  for (iter->Seek(prefix_key); iter->Valid() && iter->key().starts_with(prefix_key); iter->Next()) {
+    uint64_t expire = 0;
+    auto value = iter->value().ToString();
+    if (!decodeExpireFromValue(metadata, &value, expire).ok()) {
+      expired_field_num += 1;
+    }
+  }
+  return metadata.size - expired_field_num;
+}
+
+rocksdb::Status Hash::decodeExpireFromValue(const HashMetadata &metadata, std::string *value, uint64_t &expire) {
+  if (!metadata.IsFieldExpirationEnabled()) {
+    return rocksdb::Status::OK();
+  }
+  rocksdb::Slice data(value->data(), value->size());
+  GetFixed64(&data, &expire);
+  *value = data.ToString();
+  return (expire == 0 || expire > util::GetTimeStampMS()) ? rocksdb::Status::OK() : rocksdb::Status::NotFound();
+}
+
+rocksdb::Status Hash::encodeExpireToValue(std::string *value, uint64_t expire) {
+  std::string buf;
+  PutFixed64(&buf, expire);
+  buf.append(*value);
+  value->assign(buf.data(), buf.size());
+  return rocksdb::Status::OK();
+}
```

Comment:
```suggestion
rocksdb::Status Hash::encodeExpireToValue(std::string_view value, uint64_t expire, std::string *out) {
  PutFixed64(out, expire);
  out->append(value);
  return rocksdb::Status::OK();
}
```

Also for all places that uses this function.

Reply:
@PragmaTwice I have trouble understanding the suggestion.. would be this correct:
```
rocksdb::Status Hash::encodeExpireAndValue(std::string *value, uint64_t expire, std::string *out) {
  PutFixed64(out, expire);
  out->append(*value);
  return rocksdb::Status::OK();
}
```

Reply:
I think `string_view` should be better here.

Reply:
@PragmaTwice please let me know how 3a726e1097d665cbc7f3ee3d66d20a41fcadfdee looks, happy to address further if needed.

Reply:
@PragmaTwice string_view comment addressed via  b10d7100e24fd9c1aca35d9261736089215116e8

Reply:
Ahh I revisited it and I think it can be `std::string Hash::encodeExpireAndValue(std::string_view value, uint64_t expire)`.

The returned status here seems useless.

Reply:
```c++
std::string Hash::encodeExpireAndValue(std::string_view value, uint64_t expire) {
  std::string out;
  PutFixed64(&out, expire);
  out.append(value);
  return out;
}
```

Reply:
thanks addressed via 9bbc8d942f00ad02bbef177a685ae4e9376ede5e